src/Text/Pandoc/Format.hs

```diff
@@ -0,0 +1,118 @@
+  , disableExtension
+  , enableExtension
+  , extensionEnabled
+  , getAllExtensions
+  , getDefaultExtensions
+  , readExtension
+  )
+import Text.Parsec
+import qualified Data.Text as T
+
+-- | Format specifier with the format's name and the lists of extensions
+-- to be enabled or disabled.
+data Spec = Spec
+  { specFormat     :: T.Text
+  , specExtsDiff   :: ExtensionsDiff
+  }
+
+-- | Changes to a set of extensions, i.e., list of extensions to be
+-- enabled or disabled.
+data ExtensionsDiff = ExtensionsDiff
+  { extsToEnable  :: [Extension]
+  , extsToDisable :: [Extension]
+  }
+
+-- | Describes the properties of a format.
+data ExtensionsConfig = ExtensionsConfig
+  { extsDefault   :: Extensions -- ^ Extensions enabled by default
+  , extsSupported :: Extensions -- ^ Extensions that can be enabled or disabled.
+  }
+
+-- | Returns the extensions configuration of a format.
+getExtensionsConfig :: T.Text -> ExtensionsConfig
+getExtensionsConfig fmt = ExtensionsConfig
+  { extsDefault = getDefaultExtensions fmt
+  , extsSupported = getAllExtensions fmt
+  }
+
+-- | Apply the extension changes in the format spec to the extensions
+-- given in the format's extensions configuration. Throws an error in
+-- case of an unknown or unsupported extension.
+applyExtensionsDiff :: PandocMonad m
+                    => ExtensionsConfig
+                    -> Spec
+                    -> m Extensions
+applyExtensionsDiff extConf (Spec fname extsDiff) = do
+  let unsupported =
+        filter (\ext -> not $ extensionEnabled ext (extsSupported extConf))
+               (extsToEnable extsDiff ++ extsToDisable extsDiff)
+  case unsupported of
+    ext:_ -> throwError $ PandocUnsupportedExtensionError
+             (T.drop 4 . T.pack $ show ext) fname
+    []    -> let enabled = foldr enableExtension
+                                 (extsDefault extConf)
+                                 (extsToEnable extsDiff)
+             in pure $ foldr disableExtension enabled (extsToDisable extsDiff)
+
+-- | Parse a format-specifying string into a markup format and the
+-- change set to the format's extensions. Throws an error if the spec
+-- cannot be parsed or contains an unknown extension.
+parseFormatSpec :: PandocMonad m
+                => T.Text
+                -> m Spec
+parseFormatSpec spec =
+  case parse (fixSourcePos *> formatSpec) "" spec' of
+    Right (fname, extsDiff) -> pure (Spec (prefix <> fname) extsDiff)
+    Left err -> throwError $ PandocFormatError spec (T.pack $ show err)
+  where
+    fixSourcePos = do
+      pos <- statePos <$> getParserState
+      setPosition (incSourceColumn pos (T.length prefix))
+    formatSpec = do
+      name <- parseFormatName
+      extsDiff <- pExtensionsDiff
+      return ( T.pack name, extsDiff )
+    parseFormatName = many1 $ noneOf "-+"
+    (prefix, spec') = case T.span (/= '.') spec of
+                        (_, "") -> ("", T.toLower spec) -- no extension
+                        ps      -> ps
+
```

Comment:
The goal is to do all format processing in here, to make partial processing in [Templates](https://github.com/jgm/pandoc/blob/master/src/Text/Pandoc/Templates.hs#L85), [App.OutputSettings](https://github.com/jgm/pandoc/blob/master/src/Text/Pandoc/App/OutputSettings.hs#L91-L93), and [App](https://github.com/jgm/pandoc/blob/master/src/Text/Pandoc/App.hs#L151). This should help to fix #8137.

Custom writers paths can contain `-` and `+`, and I'm trying to account for this by splitting of the file's basename before parsing. Using the example from #8137, a path like `latex-foo-bar.lua` can currently not be parsed as a format, but allowing that will enable users to add default templates as `data/templates/default-latex-foo-bar.lua`. I figured that splitting off the basename was the easiest (and reasonably performant) way to do this. I also wanted to avoid backtracking as it might happen with paths like `latex-smart-citations.lua`. But I agree that the code is awkward, there might be a simpler way.

Reply:
Thanks for the explanation. If you want to split off the basename first, it probably makes sense to use `takeBaseName` from FilePath (even though that's for String), because note: `takeBaseName "hi.there.lua" == "hi.there"`, i.e. base names can contain a `.` too.  Alternatively, use text functions to find the *last* `.`.  And maybe add a comment about `prefix`.

Reply:
Oh true, I'll fix that.